src/drivers/amd/renoir/acp_dmic_dma.c

```diff
@@ -0,0 +1,378 @@
+	channel->direction = config->direction;
+	switch (config->direction) {
+	case DMA_DIR_DEV_TO_MEM:
+	case DMA_DIR_MEM_TO_DEV:
+		ring_buff_addr = config->elem_array.elems[0].dest &
+						DRAM_ADDRESS_MASK;
+		/* Load Ring buffer address */
+		dmic_ringbuff_addr.bits.rx_ringbufaddr = ring_buff_addr;
+		io_reg_write(PU_REGISTER_BASE +
+			ACP_WOV_RX_RINGBUFADDR, dmic_ringbuff_addr.u32all);
+		/* Load Ring buffer Size */
+		dmic_rngbuff_size = (config->elem_array.elems[0].size*
+					config->elem_array.count);
+
+		dmic_ringbuff_size.bits.rx_ringbufsize = dmic_rngbuff_size;
+		io_reg_write(PU_REGISTER_BASE +
+			ACP_WOV_RX_RINGBUFSIZE, dmic_ringbuff_size.u32all);
+		/* Write the ring buffer size to register */
+		watermark.bits.rx_intr_watermark_size = dmic_rngbuff_size/2;
+		io_reg_write(PU_REGISTER_BASE +
+			ACP_WOV_RX_INTR_WATERMARK_SIZE, watermark.u32all);
+		break;
+	default:
+		tr_err(&acp_dmic_dma_tr,
+			"dmic dma_set_config() unsupported config direction");
+		return 0;
+	}
+	if (!config->cyclic) {
+		tr_err(&acp_dmic_dma_tr,
+			"DMIC DMA: cyclic configurations are not supported!");
+		return -EINVAL;
+	}
+	if (config->scatter) {
+		tr_err(&acp_dmic_dma_tr,
+			"DMIC DMA: scatter enabled, not supported for now!");
+		return -EINVAL;
+	}
+	return 0;
+}
+
+static int acp_dmic_dma_pm_context_restore(struct dma *dma)
+{
+	//TODO
+	return 0;
+}
+
+static int acp_dmic_dma_pm_context_store(struct dma *dma)
+{
+	//TODO
+	return 0;
+}
+
+static int acp_dmic_dma_copy(struct dma_chan_data *channel, int bytes,
+			  uint32_t flags)
+{
+	struct dma_cb_data next = {
+		.channel = channel,
+		.elem.size = bytes,
+	};
+	notifier_event(channel, NOTIFIER_ID_DMA_COPY,
+			NOTIFIER_TARGET_CORE_LOCAL, &next, sizeof(next));
+	return 0;
+}
+
+static int acp_dmic_dma_probe(struct dma *dma)
+{
+	int channel;
+
+	if (dma->chan) {
+		tr_err(&acp_dmic_dma_tr, "ACP_DMIC_DMA: Repeated probe");
+		return -EEXIST;
+	}
+	tr_dbg(&acp_dmic_dma_tr, "ACP_DMIC_DMA: probe");
+	dma->chan = rzalloc(SOF_MEM_ZONE_SYS_RUNTIME, 0,
+			    SOF_MEM_CAPS_RAM, dma->plat_data.channels *
+			    sizeof(struct dma_chan_data));
+	if (!dma->chan) {
+		tr_err(&acp_dmic_dma_tr,
+			"ACP_DMIC_DMA:unable to allocate channel descriptors");
+		return -ENOMEM;
+	}
+	for (channel = 0; channel < dma->plat_data.channels; channel++) {
+		dma->chan[channel].dma = dma;
+		dma->chan[channel].index = channel;
+		dma->chan[channel].status = COMP_STATE_INIT;
+	}
+	atomic_init(&dma->num_channels_busy, 0); // TODO: Check
+	return 0;
+}
+
+static int acp_dmic_dma_remove(struct dma *dma)
+{
+	if (!dma->chan) {
+		tr_err(&acp_dmic_dma_tr,
+			"ACP_DMIC_DMA:remove called without probe");
+		return 0;
+	}
+	rfree(dma->chan);
+	dma->chan = NULL;
+	return 0;
+}
+
+static int acp_dmic_dma_get_data_size(struct dma_chan_data *channel,
+				   uint32_t *avail, uint32_t *free)
+{
+	if (channel->direction == DMA_DIR_DEV_TO_MEM) {
+		*avail = dmic_rngbuff_size/2;
+		*free = dmic_rngbuff_size/2;
```

Comment:
that seems wrong?

Usually avail+free=buffer size, but why would avail and free be equal?

Reply:
This is at the half buffer water mark, so available and free would be equall.

Reply:
No, this function is supposed to provide the actual number of bytes with data and the number of bytes free.
These numbers will vary dynamically, there is no reason why they should be pegged to the half-buffer value.
This is NOT a watermark.

Reply:
btw avail and free are the same concept in ALSA, see snd_pcm_status()

Reply:
This function provides number of bytes of data available and free in the dai buffer(DSP memory buffer) for the DMIC capture use case, not the data available or free in the host buffer (system memory). 
This data is collected based on the DMIC data copied in DSP memory.

Reply:
So SOF uses the same concepts as ALSA for both host side and DSP side DAI endpoint buffers. This is probably working today if your buffer is large enough or your DMIC DMA has divided the buffer into two periods (a ping-pong buffer). You may need to update this in the future to support different use cases.